bd-logger/src/async_log_buffer/feature_flags.rs

```diff
@@ -0,0 +1,182 @@
+/// to avoid blocking operations during construction.
+pub enum FeatureFlagsHolder {
+  /// Feature flags have not been initialized yet, only the builder is available.
+  Pending(FeatureFlagsBuilder),
+  /// Feature flags have been initialized. Contains `None` if initialization failed.
+  Initialized(Option<bd_feature_flags::FeatureFlags>),
+}
+
+impl FeatureFlagsHolder {
+  /// Creates a new holder in the pending state with the given builder.
+  ///
+  /// Feature flags will not be initialized until the first call to a method that requires them.
+  pub fn new(builder: FeatureFlagsBuilder) -> Self {
+    Self::Pending(builder)
+  }
+
+  /// Returns a reference to the feature flags if they have been initialized.
+  ///
+  /// Returns `None` if the flags are still pending or if initialization failed.
+  pub fn get(&self) -> Option<&bd_feature_flags::FeatureFlags> {
+    match self {
+      Self::Pending(_) => None,
+      Self::Initialized(flags_option) => flags_option.as_ref(),
+    }
+  }
+
+  /// Sets a feature flag to the specified variant.
+  ///
+  /// This will trigger initialization if the flags are still pending. If initialization fails,
+  /// this operation will be a no-op.
+  pub async fn set(&mut self, flag: String, variant: Option<&str>) -> LogLine {
+    if let Some(feature_flags) = self.maybe_initialize_feature_flags().await {
+      feature_flags
+        .set(flag.clone(), variant)
+        .unwrap_or_else(|e| {
+          log::warn!("failed to set feature flag: {e}");
+        });
+    }
+
+    // TODO(snowp): For all these calls we need to make sure that the timestamp on the long matches
+    // the timestamp on the feature flag update.
+
+    LogLine::new_with_fields(
+      log_level::INFO,
+      LogType::Device,
+      "Set feature flag".into(),
+      [
+        ("_set_flag".into(), AnnotatedLogField::new_ootb(flag)),
+        (
+          "_set_variant".into(),
+          AnnotatedLogField::new_ootb(variant.unwrap_or("none").to_string()),
+        ),
+      ]
+      .into(),
+    )
+  }
+
+  /// Sets multiple feature flags at once.
+  ///
+  /// This will trigger initialization if the flags are still pending. If initialization fails,
+  /// this operation will be a no-op.
+  pub async fn set_multiple(&mut self, flags: Vec<(String, Option<String>)>) -> LogLine {
+    if let Some(feature_flags) = self.maybe_initialize_feature_flags().await {
+      feature_flags
+        .set_multiple(flags.clone())
+        .unwrap_or_else(|e| {
+          log::warn!("failed to set feature flags: {e}");
+        });
+    }
+
+    LogLine::new_with_fields(
+      log_level::INFO,
+      LogType::Device,
+      "Set multiple feature flags".into(),
+      flags
+        .into_iter()
+        .map(|(flag, variant)| {
+          (
+            format!("_set_flag_{flag}").into(),
+            AnnotatedLogField::new_ootb(variant.unwrap_or_else(|| "none".to_string())),
+          )
+        })
+        .collect(),
+    )
```

Comment:
If we add a new log type, IMO we should use a binary format for efficiency (similar to session replay). Also, not sure if we should consider chunking super large batch operations? Like what if we have the 2K set case, should we send 2K in the single log line? It could get pretty big.

The downside of course of the binary format is that it makes matching much more difficult. I wonder if we should keep some dedicated map for matching, and then before it's serialized the buffer convert this to binary format? the on demand serialization would also be required for field compression so possibly some shared machinery there?